### DISCUSSION TITLE
ros2 topic list does not display the robot's topics using ethernet over USB connection

### DISCUSSION THREAD

**Discussion by AlHussain00** [iRobotEducation/create3_docs#234] [Questions] [answered]
Hello everyone,

I am connecting to the robot via **ethernet over usb** connection, and was able to access the web server to update the firmware as recommended in the documentation. Moreover, I made sure that the switch on the board is toggled to USB. When I use **ros2 topic list** nothing shows except for _/parameter_events_ and _/rosout_.

I am using ubuntu 20.04 on my PC, ROS2 Galactic, and CycloneDDS. In addition, I made sure that the robot has CycloneDDS configuration from the web server. I also tried to set up xml profile and specify my nic, as shown below:
```
<CycloneDDS>
   <Domain>
     <General>
                       <NetworkInterfaceAddress>NETWORKINTERFACEHERE</NetworkInterfaceAddress>
    </General>
   </Domain>
</CycloneDDS>
```
then I exported this xml file as follows: 

`export CYCLONEDDS_URI=/path/to/the/xml/profile.xml`

I even checked the network by using the following command:

`ping 192.168.186.2`

and got the following output:

```
PING 192.168.186.2 (192.168.186.2) 56(84) bytes of data.
64 bytes from 192.168.186.2: icmp_seq=1 ttl=64 time=0.624 ms
64 bytes from 192.168.186.2: icmp_seq=2 ttl=64 time=0.659 ms
64 bytes from 192.168.186.2: icmp_seq=3 ttl=64 time=0.583 ms
64 bytes from 192.168.186.2: icmp_seq=4 ttl=64 time=0.653 ms
^C
--- 192.168.186.2 ping statistics ---
4 packets transmitted, 4 received, 0% packet loss, time 3064ms
rtt min/avg/max/mdev = 0.583/0.629/0.659/0.030 ms
```
which shows that the network between my PC and the robot is fine. (correct me if I am mistaken)

Finally, I ran the following commands:

`ros2 daemon stop`
`ros2 topic list
`

However, the problem remained the same. I ran out of ideas, could you please help me out?

Thank you in advance.

Regards,

**Comment by alsora** [comment]
Is the Create 3 also connected to the WiFi?
If yes, you are running with "multiple network interfaces" which requires special configuration.
See https://iroboteducation.github.io/create3_docs/setup/xml-config/#multiple-network-interfaces_1.
In particular, you may try to add the `<DontRoute>true</DontRoute>` argument to the XML configuration as shown there.

As an alternative, you may try to switch both the create 3 and the laptop to use Fast-DDS.
This DDS implementation should work out of the box without need to set XML configuration files.

**Reply by AlHussain00** [reply]
No, Create 3 is not connected to the WiFi. I even tried to use Fast-DDS configuration instead of CycloneDDS, but the issue was not solved.

**Reply by justinIRBT** [reply]
Can you show your `ifconfig` on your laptop for your ethernet over usb address?

**Reply by AlHussain00** [reply]
I ran `ifconfig` and got the following:

```
enp3s0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 131.234.250.78  netmask 255.255.254.0  broadcast 131.234.251.255
        inet6 2001:638:502:250:2701:a12b:4503:6a74  prefixlen 64  scopeid 0x0<global>
        inet6 fe80::2f41:673f:9fd0:84b4  prefixlen 64  scopeid 0x20<link>
        inet6 2001:638:502:250:b67a:9c23:b289:88aa  prefixlen 64  scopeid 0x0<global>
        ether 00:d8:61:5c:82:d9  txqueuelen 1000  (Ethernet)
        RX packets 10568  bytes 5465816 (5.4 MB)
        RX errors 0  dropped 763  overruns 0  frame 0
        TX packets 3805  bytes 932313 (932.3 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

enx503f56024083: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.186.1  netmask 255.255.255.0  broadcast 192.168.186.255
        inet6 fe80::ade:6c4e:640f:7b22  prefixlen 64  scopeid 0x20<link>
        ether 50:3f:56:02:40:83  txqueuelen 1000  (Ethernet)
        RX packets 218  bytes 94495 (94.4 KB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 360  bytes 48456 (48.4 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 1348  bytes 147947 (147.9 KB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 1348  bytes 147947 (147.9 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

where **enx503f56024083** is the nic name.

**Comment by AlHussain00** [comment, marked as answer]
After one week of frustration, I was able to figure out why Create 3 was not working. It turned out that the order of the following lines matters within _bashrc_ file:

```
source /usr/share/colcon_cd/function/colcon_cd.sh
export _colcon_cd_root=/opt/ros/galactic/
source /usr/share/colcon_argcomplete/hook/colcon-argcomplete.bash
source /opt/ros/galactic/setup.bash
export RMW_IMPLEMENTATION=rmw_fastrtps_cpp
```
I also switch the configuration to Fast-DDS as indicated above. However, I am not able to tell why this is important, and it will be great if there is an explanation for that.

I would like to thank everyone, who tried to help me.

Regards,

**Reply by justinIRBT** [reply]
I haven't personally used the colcon related `.sh` files, but it surprises me to hear that the order mattered, I haven't found any order dependencies with the `galactic/setup.bash` and the `export RMW_IMPLEMENTATION`.  I would suspect switching to Fast-DDS made a difference.  Fast-DDS is a little more robust to various network configurations out of the box (for example, it automatically runs on all interfaces.)  Cyclone requires the right config for things to be happy.  If you specified `enx503f56024083` as your network name in the cycloneDDS XML file, I would have expected that to work, but maybe there is something underlying about your network setup that CycloneDDS is more sensitive to.

**Reply by alsora** [reply]
+1 on what Justin said.
The order of all these environment variables is not important.
They all have different effects that do not interfere with each others.

Your first 3 lines are to setup the colcon cd and arg completion tools.
They have no impact on running ROS 2 applications.
```
source /usr/share/colcon_cd/function/colcon_cd.sh
export _colcon_cd_root=/opt/ros/galactic/
source /usr/share/colcon_argcomplete/hook/colcon-argcomplete.bash
```

The 4th command sets environment variables such as `PATH`, `PYTHONPATH` and `LD_LIBRARY_PATH` pointing at your ROS 2 galactic environment.
This command is needed before building and running executables.
Trying to `ros2 run ...` an application or calling `ros2 topic ...` without first sourcing the environment will result in a "command not found" error.
```
source /opt/ros/galactic/setup.bash
```

The last command sets the RMW implementation.
This command is evaluated when you start a ROS 2 application, which will parse this environment variable when creating a ROS 2 context.
```
export RMW_IMPLEMENTATION=rmw_fastrtps_cpp
```

However, note that the `RMW_IMPLEMENTATION` can be cached by the ROS 2 daemon, that's why in the documentation we remind to run with `--no-daemon` to avoid this type of issues.
https://iroboteducation.github.io/create3_docs/examples/sensors-cli/#overview

It's likely that between the different tests you ended up resetting the ROS 2 daemon (either voluntarily or not)
